pkg/controller/nodes/executor_test.go

```diff
@@ -963,6 +963,69 @@ func TestNodeExecutor_RecursiveNodeHandler_Recurse(t *testing.T) {
 		assert.Equal(t, uint32(0), mockNodeStatus.GetAttempts())
 		assert.Equal(t, v1alpha1.NodePhaseFailing.String(), mockNodeStatus.GetPhase().String())
 	})
+
+	// not fail immediately for last retry failure when enabled cleanupLastRetry
+	t.Run("not-fail-immediately-for-last-failure", func(t *testing.T) {
+		hf := &mocks2.HandlerFactory{}
+		store := createInmemoryDataStore(t, promutils.NewTestScope())
+		adminClient := launchplan.NewFailFastLaunchPlanExecutor()
+		execIface, err := NewExecutor(ctx, config.GetConfig().NodeConfig, store, enQWf, mockEventSink, adminClient,
+			adminClient, 10, "s3://bucket", fakeKubeClient, catalogClient, promutils.NewTestScope())
+		assert.NoError(t, err)
+		exec := execIface.(*nodeExecutor)
+		exec.cleanupLastRetry = true
+		exec.nodeHandlerFactory = hf
+
+		h := &nodeHandlerMocks.Node{}
+		h.On("Handle",
```

Comment:
you can start using the new interface for writing mock handlers, i.e., in this case
```
h.OnHandle(....).Return(...)
```
It becomes simpler to refactor